src/FeedGenerator.php

```diff
@@ -561,4 +533,42 @@ protected function get_batch_size(): int {
 	protected function process_item( $item, array $args ) {
 		// Process each item here.
 	}
+
+	/**
+	 * Return the list of supported product types.
+	 *
+	 * @since x.x.x
+	 *
+	 * @return array
+	 */
+	private function get_supported_product_types(): array {
+		return (array) apply_filters(
+			'pinterest_for_woocommerce_included_product_types',
+			array(
```

Comment:
Given that this is a similar list to the `AttributesTab` class, it might make sense to abstract this into a `trait` that can be included in both classes.

Reply:
Just a couple of comments before going with this approach.

1. @JPry, we are not doing any diff operation with this array. Are you suggesting to use a `trait` and set the filter name and the default array as class parameters, or to use a key-based array on this method?

2. @jconroy with these changes we are including in the feed only the supported products listed here (simple, external, variation). The customer can add extra product types to the feed filtering that value with a php snippet. This also will break existing installs with custom product types. Just wanna add this extra confirmation, since some customers may be not too happy with using php to add new product types to the feed. What do you think?

Reply:
@alexiglesias31 

> we are not doing any diff operation with this array. Are you suggesting to use a `trait` and set the filter name and the default array as class parameters, or to use a key-based array on this method?

Sorry that I wasn't more clear in my comment. It looked to me like it was a similar set of product types between this class and the `AttributesTab` class. The only difference is that this class includes the `external` product type.

My thinking was that this might be a place for combining the two arrays into one location. But with that said, I don't feel very strongly about it, so if it doesn't make sense then we can skip it. I wasn't sure where else these kinds of things might be needed in the future.

I'll let @jconroy or @budzanowski comment on point 2, but if we have documentation in place around the use of the filter, then that should help users have proper expectations around what changing this value will impact. We could also ensure that the place where we pull in the product types can handle custom types gracefully. 

We should also try to handle the deprecated `pinterest_for_woocommerce_excluded_product_types` filter with a proper warning, and if at all possible, an option to use the old behavior of that filter.

Reply:
> This also will break existing installs with custom product types.

We should have rectified the use of that filter a long time ago when the user base was very small. I think that it is still small enough that we should try to do that. Maybe the fallout will not be big and as @JPry mentions with proper documentation we should minimise the potential hassle after the release. We would need to warn the support people about this beforehand. My guess is that this should be rather painless ( #famouslastwords ) - maybe we will even fix some of the scenarios like https://github.com/woocommerce/pinterest-for-woocommerce/issues/543

> We should also try to handle the deprecated pinterest_for_woocommerce_excluded_product_types filter with a proper warning, and if at all possible, an option to use the old behaviour of that filter.

I agree with that. I think that we can do this with small change to the code organisation: 

```PHP
			/**
			 * Filter excluded product types.
			 *
			 * @since 1.0.10
			 * @param array Array of excluded product types.
			 */
			$excluded_product_types = apply_filters(
				'pinterest_for_woocommerce_excluded_product_types',
				array(
					'grouped',
					'variable',
					'subscription',
					'variable-subscription',
				)
			);

		   $supported_product_types = self::get_supported_product_types();

           $included_product_types = array_diff( $supported_product_types, $excluded_product_types );
```

The construction looks a bit over-engineered but it preserves the old behaviour of filtering out products using 'pinterest_for_woocommerce_excluded_product_types' and with the new filter new product types could be added. 

Ultimately the plugin that extends the list of available product types should declare using the `pinterest_for_woocommerce_included_product_types` filter that it supports Pinterest integration and the product makes sense in the Feed file.

@jconroy, @JPry, @alexiglesias31 thoughts?

Reply:
@budzanowski I'm totally agree with your comments.
I think allowing only certain defined product types in Google's feed and terms is the way to go. My main concern about this (and the cause of my warning comment :slightly_smiling_face: ) is users reaction. But to be honest, I think this will save us some headaches in the future, as any kind of custom product can break the feed or be displayed incorrectly.

Reply:
@alexiglesias31 great. I will let others to chime in on what we are proposing here. @jconroy maybe there are other people we should ask for opinion?

For the meantime please adjust the code preserve the `pinterest_for_woocommerce_excluded_product_types` functionality. We will need to prepare documentation update bc this filter is not covered anywhere.

Reply:
> Ultimately the plugin that extends the list of available product types should declare using the pinterest_for_woocommerce_included_product_types filter that it supports Pinterest integration and the product makes sense in the Feed file.

Agree

>  I think this will save us some headaches in the future, as any kind of custom product can break the feed or be displayed incorrectly.

Agree.

> We would need to warn the support people about this beforehand. My guess is that this should be rather painless ( #famouslastwords ) - maybe we will even fix some of the scenarios like https://github.com/woocommerce/pinterest-for-woocommerce/issues/543

Lets chat with Sunday on his recommendation for best way to do this. 

Before we ship I think it would be good to give HEs a heads up and [add a section to docs like this one about what products types are supported by default](https://woocommerce.com/document/google-listings-and-ads-faqs/#section-3).

Agree it would be nice to add a warning and give option to use old handling if they want.

I think the fact that we've been included product types that we generally know aren't supported (subscriptions/composite products) I agree it should be better overall for customers.

Reply:
@JPry

> Sorry that I wasn't more clear in my comment. It looked to me like it was a similar set of product types between this class and the AttributesTab class. The only difference is that this class includes the external product type.
My thinking was that this might be a place for combining the two arrays into one location. But with that said, I don't feel very strongly about it, so if it doesn't make sense then we can skip it. I wasn't sure where else these kinds of things might be needed in the future.

Giving a second though to this :thinking:, perhaps external product should be included/excluded in both places. Any thoughts @jconroy? I didn't find external products on GLA extension.

Reply:
@budzanowski 

> The construction looks a bit over-engineered but it preserves the old behaviour of filtering out products using pinterest_for_woocommerce_excluded_product_types and with the new filter new product types could be added.

The filter is back :ok_hand:. Not sure if we should exclude composite products by default too :thinking:.
cc @jconroy

Reply:
Google merchant center [doesn't allow affiliate products](https://support.google.com/merchants/answer/6363310?hl=en) so [external aren't supported](https://github.com/woocommerce/google-listings-and-ads/blob/e17daaf1526493eeac1febce6b8d60f57b9ffe0f/src/Product/ProductSyncer.php#L256).

I'd follow the same approach for Pinterest